File: electron/window/handler/cloudflared.js
const { ipcMain } = require("electron");
const { exec } = require("child_process");

// const {} = require("../../../config/cloudflared.json");

/**
 * @returns {Promise<import("./cloudflared").STATES>}
 */
async function getState() {
    return new Promise((resolve, reject) => {
        exec(`sc query cloudflared`, (error, stdout, stderr) => {
            if (error) reject(error);
            else {
                const split = stdout
                    .split("\n")
                    .map((line) => line.trim())
                    .filter((line) => line.length);
                const stateLine = split.find((line) => line.startsWith("STATE"));
                const splitLine = stateLine.split(" ").filter((line) => line.length);
                const state = splitLine[splitLine.length - 1];
                resolve(state);
            }
        });
    });
}

async function toggleTunnel(enable) {
    await new Promise((resolve, reject) => {
        const action = enable ? "start" : "stop";
        exec(`powershell -Command "Start-Process cmd -Verb RunAs -ArgumentList '/c sc ${action} cloudflared' -WindowStyle Hidden"`, (error) => {
            if (error) reject(error);
            else resolve();
        });
    });
}

ipcMain.handle("getState", async () => {
    // console.log("getState");
    return getState();
});

ipcMain.handle("toggleTunnel", async (event, enable) => {
    // console.log("toggleTunnel", enable);
    return toggleTunnel(enable)
        .then(() => true)
        .catch(() => false);
});

File: src-tauri/src/commands/cloudflared.rs
use std::{os::windows::process::CommandExt, process::Command};

#[derive(serde::Serialize)]
pub enum Status {
    RUNNING,
    STOPPED,
    ERROR,
    UNKNOWN,
}

impl From<&str> for Status {
    fn from(status: &str) -> Self {
        match status {
            "RUNNING" => Status::RUNNING,
            "STOPPED" => Status::STOPPED,
            "ERROR" => Status::ERROR,
            _ => Status::UNKNOWN,
        }
    }
}

#[tauri::command]
pub fn cloudflared_toggle(enable: bool) -> Result<(), String> {
    let status = if enable { "start" } else { "stop" };
    let arg: String = format!(
        "Start-Process cmd -Verb RunAs -ArgumentList '/c sc {} cloudflared' -WindowStyle Hidden",
        status
    );

    let output = Command::new("powershell")
        .arg("-Command")
        .arg(arg)
        .creation_flags(0x08000000) // CREATE_NO_WINDOW
        .output()
        .map_err(|e| format!("Failed to execute cloudflared: {}", e))?;

    if !output.status.success() {
        return Err(format!(
            "cloudflared {} failed: {}",
            status,
            String::from_utf8_lossy(&output.stderr)
        ));
    }

    Ok(())
}

// #[tauri::command(rename = "cloudflared_status")]
#[tauri::command]
pub fn cloudflared_status() -> Result<Status, String> {
    let output = Command::new("sc")
        .arg("query")
        .arg("cloudflared")
        .creation_flags(0x08000000) // CREATE_NO_WINDOW
        .output()
        .map_err(|e| format!("Failed to query cloudflared status: {}", e))?;

    if !output.status.success() {
        return Err(format!(
            "Failed to query cloudflared status: {}",
            String::from_utf8_lossy(&output.stderr)
        ));
    }

    let string_output = String::from_utf8_lossy(&output.stdout);
    let line = string_output
        .lines()
        .find(|line| line.trim().starts_with("STATE"))
        .unwrap_or("");
    let status = line.split_ascii_whitespace().last().unwrap_or("UNKNOWN");
    println!("State: {}", status);

    Ok(Status::from(status))
}
